tsl/src/compression/recompression.c

```diff
@@ -0,0 +1,863 @@
+{
+	if (tuplestorecxt)
+	{
+		tuplesort_end(tuplestorecxt->tuplestore);
+		pfree(tuplestorecxt->sort_keys);
+		pfree(tuplestorecxt->sort_operators);
+		pfree(tuplestorecxt->sort_collations);
+		pfree(tuplestorecxt->nulls_first);
+		pfree(tuplestorecxt);
+	}
+}
+
+/*
+ * Helper method to fetch all rows from uncompressed chunk
+ * and sort these rows and save it in tuplestore
+ */
+static int
+save_uncompressed_rows_in_tuplestore(Chunk *uncompressed_chunk,
+									 Tuplesortstate *uncompressed_rows_sortstate)
+{
+	Relation uncompressed_chunk_rel;
+	TableScanDesc heapScan;
+	HeapTuple uncompressed_tuple;
+	TupleTableSlot *heap_tuple_slot = NULL;
+	int total_uncompressed_rows = 0;
+
+	uncompressed_chunk_rel = RelationIdGetRelation(uncompressed_chunk->table_id);
+	heapScan = table_beginscan(uncompressed_chunk_rel, GetLatestSnapshot(), 0, NULL);
+	heap_tuple_slot =
+		MakeTupleTableSlot(RelationGetDescr(uncompressed_chunk_rel), &TTSOpsHeapTuple);
+
+	while ((uncompressed_tuple = heap_getnext(heapScan, ForwardScanDirection)) != NULL)
+	{
+		ExecStoreHeapTuple(uncompressed_tuple, heap_tuple_slot, false);
+		slot_getallattrs(heap_tuple_slot);
+		tuplesort_puttupleslot(uncompressed_rows_sortstate, heap_tuple_slot);
+		simple_heap_delete(uncompressed_chunk_rel, &uncompressed_tuple->t_self);
+		total_uncompressed_rows++;
+	}
+	RelationClose(uncompressed_chunk_rel);
+	ExecDropSingleTupleTableSlot(heap_tuple_slot);
+	table_endscan(heapScan);
+	/* sort the tuplestore */
+	tuplesort_performsort(uncompressed_rows_sortstate);
+	return total_uncompressed_rows;
+}
+
+/*
+ * Helper method to fetch _ts_meta_count from a compressed tuple.
+ */
+static void
+fetch_segment_metadata_count(RowDecompressor *decompressor, HeapTuple compressed_tuple,
+							 Datum *metacount_val)
+{
+	Datum *values = palloc0(sizeof(Datum) * decompressor->in_desc->natts);
+	bool *nulls = palloc0(sizeof(bool) * decompressor->in_desc->natts);
+	heap_deform_tuple(compressed_tuple, decompressor->in_desc, values, nulls);
+	if (metacount_val)
+	{
+		/* fetch number of rows present in current segment of compressed tuple */
+		AttrNumber meta_count =
+			get_attnum(decompressor->in_rel->rd_id, COMPRESSION_COLUMN_METADATA_COUNT_NAME);
+		*metacount_val = values[meta_count - 1];
+	}
+	pfree(values);
+	pfree(nulls);
+}
+
+/*
+ * Helper method to get total rows per segment before recompressing
+ * that particular segment. This helps us decide if we need to split
+ * a segment into 2 equal halfs or not.
+ */
+static unsigned long
+get_total_rows_per_segment(RowDecompressor *decompressor, HeapTuple compressed_tuple,
+						   unsigned long uncompressed_rows_per_segment)
+{
+	Datum meta_count = 0;
+	fetch_segment_metadata_count(decompressor, compressed_tuple, &meta_count);
+	if ((meta_count + uncompressed_rows_per_segment) >= MAX_ROWS_PER_COMPRESSION)
+		return (meta_count + uncompressed_rows_per_segment);
+	else
+		return (2 * MAX_ROWS_PER_COMPRESSION);
```

Comment:
Can we just avoid this kind of arIthmetics at all? If we have such code, we have to think hard to prove that the actual limit is still `MAX_ROWS_PER_COMPRESSION`.


I can imagine roughly the following way to deal with overflowing batches:

We have to incorporate new rows into the already compressed data. To understand what work is required, we will track the ranges of compressed data that are affected by the recompression. Compressed data is ordered by (segmentby, orderby) values. One range is an interval on this axis and the respective metadata. 

First, go over the uncompressed rows ordered by (segmentby, orderby).
Keep track of the current affected compressed range:
	- segmentby
	- min (orderby)
	- max (orderby)
	- row count
	- starting sequence number
	- next adjacent sequence number (of the first next batch that is not affected)

For each row:
    - does it fall into the current range?
    	* if not, the current range is ready for compression! remember it and start the new one with this row.

	- find the intersecting compressed batch
		* if found, adjust the range to include it

	- check that the current compressed range is re-compressible by itself, without affecting the compressed batches: it should have enough sequence numbers to fit its count of rows as compressed batches of max size
		* if not, include the next compressed segment into the range
		* loop until true

After we built these compressed ranges, we know that each can be compressed using our normal algorithm, and we will have enough sequence numbers for it.
 
I don't like that we resort after each compressed segment, this is a lot of extra work, would be good to find a way to only re-sort once. Maybe we could put the decompressed batches into a second tuplesort.

Do you think we could refactor it along these lines?